skfp/utils/splitters.py

```diff
@@ -0,0 +1,221 @@
+        The fraction of data to be used for the validation split.
+
+    frac_test : float, default=0.1
+        The fraction of data to be used for the test split.
+
+    return_indices : bool, default=False
+        Should the method return SMILES/RDKit Mol objects or just indices in the dataset.
+
+    include_chirality: bool, default=False
+        Should the method take chirality of molecules into consideration?
+
+
+    Returns
+    ----------
+    subsets : tuple[list, list, list]
+        A tuple of train, validation, and test subsets. The format depends on `return_indices`:
+        - If `return_indices` is False, returns lists of SMILES strings
+        or RDKit Mol objects depending on the input.
+        - If `return_indices` is True, returns lists of indices.
+
+    Notes
+    -----
+    - This implementation uses Bemis-Murcko scaffolds (Bemis and Murcko, 1996) to group molecules.
+    - The split is fully deterministic, with the smallest scaffold sets assigned to the test subset,
+    larger sets to the validation subset,
+      and the largest sets to the training subset.
+    - The test and validation sets can be slightly larger than specified,
+    ensuring they are at least the desired fraction, but as close as possible in size.
+    - The split fractions (frac_train, frac_valid, frac_test) must sum to 1.
+
+    References
+    ----------
+    - Bemis, G. W., & Murcko, M. A. (1996). The properties of known drugs. 1. Molecular frameworks.
+    Journal of Medicinal Chemistry, 39(15), 2887-2893.
+    - Z. Wu, B. Ramsundar, E. N. Feinberg, J. Gomes, C. Geniesse, A. S. Pappu, K. Leswing, V. Pande (2017).
+    MoleculeNet: A Benchmark for Molecular Machine Learning. Chemical Science, 9(2), 513-530.
+    """
+    np.testing.assert_almost_equal(train_size + test_size + valid_size, 1.0)
+
+    scaffolds = create_scaffolds(list(data), include_chirality)
+    scaffold_sets = sorted(scaffolds.values(), key=len)
+
+    train_cutoff: int = int(train_size * len(data))
+    valid_cutoff: int = int((train_size + valid_size) * len(data))
+
+    train_ids: list[int] = []
+    valid_ids: list[int] = []
+    test_ids: list[int] = []
+
+    for scaffold_set in scaffold_sets:
+        if len(train_ids) < train_cutoff:
+            train_ids.extend(scaffold_set)
+        elif len(valid_ids) < valid_cutoff:
+            valid_ids.extend(scaffold_set)
+        else:
+            test_ids.extend(scaffold_set)
+
+    if return_indices:
+        return train_ids, valid_ids, test_ids
+    else:
+        return (
+            _get_data_from_indices(data, scaffolds, train_ids),
+            _get_data_from_indices(data, scaffolds, valid_ids),
+            _get_data_from_indices(data, scaffolds, test_ids),
+        )
+
+
+def _get_data_from_indices(data, scaffolds, indices):
+    """
+    Helper function to retrieve data from indices.
+    """
+    return [data[i] if isinstance(data[0], str) else scaffolds[i] for i in indices]
```

Comment:
1. This will be really slow for lists. Change `indices` to a set.
2. Why `isinstance(data[0], str)`?